### PR TITLE
Keep sockets in native / wavefront

### PR DESCRIPTION
**Pull request by blt**
The previous implementation of these sinks created and destroyed sockets
on each flush. There is no good reason for this any longer. We now take
the approach of preserving the socket only so long as we're able to send
information through it, otherwise we destroy the socket and get a new one.

Signed-off-by: Brian L. Troutwine <blt@postmates.com>

### CODE REVIEW

**Comment by blt** [comment]
@bitglue relevant to your interests

**Comment by bitglue** [comment]
Is this what's causing the RST every 10 seconds? Or was that the firehose sink?

**Comment by blt** [comment]
@bitglue I _believe_ it's the RST sink, but we don't control that socket. It's hard to fix. On inspection of the codebase I realized we could fix this too.

**Comment by bitglue** [comment]
By "RST sink" you mean "Firehose sink"?

**Comment by blt** [comment]
@bitglue I meant firehose sink, yes. I have _no idea_ why I wrote "RST sink". That is gibberish.

File: src/sink/wavefront.rs [hunk 1]
```diff
@@ -317,43 +352,24 @@ impl Sink for Wavefront {
...
+                    self.aggrs.reset();
+                    self.stats.clear();
+                    self.delivery_attempts = 0;
+                    return;
+                } else {
+                    self.delivery_attempts = self.delivery_attempts.saturating_add(1);
+                    delivery_failure = true;
                 }
+            } else {
+                time::delay(self.delivery_attempts);
```
Comment (tsantero) [inline src/sink/wavefront.rs]: I can't tell where `time::delay` is coming from. Is the fundamental unit passed in seconds? Perhaps maybe we want `std::thread::sleep` instead here?

Also, it seems like it's possible for this to be hit when `self.delivery_attempts` is `0`, which I don't think is a behavior we want?

The backoff logic here is linear. Would we prefer exponential?

Comment (blt) [inline src/sink/wavefront.rs]: `time::delay` comes from [here](https://github.com/postmates/cernan/blob/master/src/time.rs#L38). The inputs are linear but the backoff occurs as the power of two, up to a cap. We special-case the `0` to be a no-op https://github.com/postmates/cernan/blob/master/src/time.rs#L40.

Comment (tsantero) [inline src/sink/wavefront.rs]: this all makes so much more sense when you actually read the function being called 😂

**Pull request merged**